tests/GitHub/Workflow/Command/AnnotationSpec.hs

```diff
@@ -0,0 +1,56 @@
+module GitHub.Workflow.Command.AnnotationSpec
+  ( spec
+  ) where
+
+import Data.Function (($), (&))
+import GitHub.Workflow.Command.Annotation
+import Test.Hspec
+
+spec :: Spec
+spec =
+  context "Annotation" do
+    specify "debug with no properties" $
+      toByteString (debug "Set the Octocat variable")
+        `shouldBe` "::debug::Set the Octocat variable"
+
+    specify "error with no properties" $
+      toByteString (error "Missing semicolon")
+        `shouldBe` "::error::Missing semicolon"
+
+    specify "error with file" $
+      toByteString (error "Missing semicolon" & setLocation (fromFile "app.js"))
+        `shouldBe` "::error file=app.js::Missing semicolon"
+
+    specify "error with line number" $
+      toByteString
+        ( error "Missing semicolon"
+            & setLocation (fromFile "app.js" & setSingleLinePosition (fromLine 1))
+        )
+        `shouldBe` "::error file=app.js,line=1::Missing semicolon"
+
+    specify "notice with column range" $
+      toByteString
+        ( notice "Missing semicolon"
+            & setLocation
+              ( fromFile "app.js"
+                  & setSingleLinePosition (fromLine 1 & setColumnRange (ColumnRange 5 7))
+              )
+        )
```

Comment:
Personally, for me, I think the types have overshot the sweet-spot here, and it comes out with more verbose usage like this.

I wonder if dialing that back a bit (have just one `HasLocation` class that's `Maybe`, have just a single constructor for things that are single-vs-multi by giving them `Maybe` end fields instead, not `newtype`-wrapping _everything_, ditching the isos, etc) could result in nicer ergonomics. Something like:

```hs
notice "Missing semicolon"
  & location ?~ ("app.js"         -- abuse IsString, take it or leave it
    & startLine ?~ (1             -- abuse Num, probably leave it :)
      & startColumn .~ 5          -- start is required, so .~
      & endColumn ?~ 7))          -- end is optional, so ?~
```

Then simple cases get simpler, with the verbosity scaling up only as necessary:

```hs
notice "Missing semicolon"
-- same as yours

notice "Missing semicolon" & location ?~ "app.js"
notice "Missing semicolon" & setLocation (fromFile "app.js") -- vs

notice "Missing semicolon" & location ?~ ("app.js" & startLine ?~ 1)
notice "Missing semicolon" & setLocation (fromFile "app.js" & setSingleLinePosition (fromLine 1)) -- vs
```

What you have already has all obviously been well thought-out, so there are likely benefits to all the typing that I'm missing, just wanted to put it out there.

Reply:
Yeah, too much.

It stemmed from a modeling mistake; since start line is always present in a position with a file, `Position` should have been a record with start line as a field, rather than duplicating that across two constructors in a sum type. What remains, then, is something that needed a name, and I'm now calling that the position's "extent", which is either `WithinLine` or `ToLine`. I think that's the only bit that's a little more verbose than you'd like.

Removed all the typeclasses that only had one interesting instance.

Removed the iso classes, just exporting constructors for newtypes that don't need to be abstract.

The API is now based on lens and doesn't try to do anything clever at all, just exposes the algebraic datatypes which makes it easier to form a mental model.